Added the possibility to compile simbody with MinGW for Windows

Gjacquenot opened the pull request:
These commits enable compilation with various versions of MinGW for Windows.

Since, simbody is shipped with Blas and Lapack libraries compiled with specific versions of MinGW, some checks had to be added to verify compatibility.

If it is possible to compile simbody, configuration and compilation run smoothly.
Otherwise, user is asked to provide its own version of Blas and Lapack, or to change to its version of MinGW.

Versions tested:

- MinGW 32 with gcc 4.7.2 and dwarf exception mechanism
- MinGW 64 with gcc 4.9.2 and SJLJ exception mechanism
- MinGW 64 with gcc 5.2.0 and SEH exception mechanism with Blas and Lapack compiled manually



Please note, that one can clean the code with the the two Python scripts:
    
The Python script that has been used to remove trailing spaces from CMakeLists.txt files found recursively is:
    
    import os
    PATH = r'simbody'
    for path, dirs, files in os.walk(PATH):
        for f in files:
            file_name, file_extension = os.path.splitext(f)
            if file_name == 'CMakeLists' and file_extension =='.txt':
                print(f)
                path_name = os.path.join(path, f)
                with open(path_name, 'r') as fh:
                    new = [line.rstrip() for line in fh]
                with open(path_name, 'w') as fh:
                    fh.writelines((line+'\n' for line in new))

The Python script that has been used to remove trailing spaces from files *.c *.h *.cpp *.hpp found recursively is:
    
    import os
    PATH = r'simbody'
    extensions = ('.c','.h','.cpp','.hpp')
    
    for path, dirs, files in os.walk(PATH):
        for f in files:
            file_name, file_extension = os.path.splitext(f)
            if file_extension in extensions:
                print(f)
                path_name = os.path.join(path, f)
                with open(path_name, 'r') as fh:
                    new = [line.rstrip() for line in fh]
                with open(path_name, 'w') as fh:
                    fh.writelines((line+'\n' for line in new))

sherm1 commented:
Hi, @Gjacquenot. Thanks very much for this -- it looks great. The README documentation and instructions are excellent. Sorry for the slow review -- we've been very busy.

I still haven't had a chance yet to test this to make sure nothing else broke -- I'll try to get to that today.

sherm1 commented:
@Gjacquenot, I tried your branch in Windows 8.1 with VC++ 2015; it worked fine and all tests passed.

Then I tried generating "MSYS Makefiles" (using the CMake 3.3.1 GUI) and got:
```
The C compiler identification is GNU 4.7.0
The CXX compiler identification is GNU 4.7.0
Check for working C compiler: c:/mingw/bin/gcc.exe
Check for working C compiler: c:/mingw/bin/gcc.exe -- works
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Detecting C compile features
Detecting C compile features - done
Check for working CXX compiler: c:/mingw/bin/g++.exe
Check for working CXX compiler: c:/mingw/bin/g++.exe -- works
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Detecting CXX compile features
Detecting CXX compile features - done
Found Doxygen: C:/Program Files/doxygen/bin/doxygen.exe (found version "1.8.8") 
MINGW_BIN_PATH = c:/mingw/bin
CMake Error at Platform/CMakeLists.txt:51 (if):
  if given arguments:

    "STREQUAL" "seh"

  Unknown arguments specified
```
That line in Platform/CMakeLists.txt is:
```
            if( ${MINGW_GCC_EXCEPTION} STREQUAL "seh" )
```
So it looks like the variable must have been empty.

Ideas?

sherm1 commented:
I did find `libgcc_s_sjlj-1.dll` in my mingw installation but it was in directory `C:/mingw/x86_64-w64-mingw32/lib`. The only things in mingw/bin were .exe files.

chrisdembia commented on README.md:
```diff
@@ -492,3 +546,4 @@ Prof. Scott Delp is the Principal Investigator on these grants and Simbody is us
 [flores]: http://xray.bmc.uu.se/flores/Home.html
 [buildwin]: https://github.com/simbody/simbody/raw/master/doc/HowToBuildSimbodyFromSource_Windows.pdf
 [buildunix]: https://github.com/simbody/simbody/raw/master/doc/HowToBuildSimbodyFromSource_MacLinux.pdf
+ bits32
```
@Gjacquenot did you intend to add this `bits32` to the readme?

chrisdembia commented on README.md:
```diff
@@ -466,6 +466,60 @@ Simbody is installed into the `usr/` directory.  The directory
+
+This build is still experimental, because of the precompiled libraries
+Simbody depends on (Blas and Lapack).
+To ease the compilation Windows, Simbody provides precompiled libraries for Blas and Lapack.
```
This sentence has a typo: "To ease the compilation Windows". Perhaps "To ease building on Windows"?

Also, "precompiled libraries" makes me think of "precompiled headers". Wouldn't it be sufficient to say "compiled libraries"?

Gjacquenot replied:
@chrisdembia : for all these typos, should I

- create a new commit for each typo
- create a new commit for all typos
- create and squash all typos in a previous commit

chrisdembia commented on README.md:
```diff
@@ -466,6 +466,60 @@ Simbody is installed into the `usr/` directory.  The directory
+* On Windows 32 Bits, these were compiled with a Dwarf exception mechanism,
+* On Windows 64 Bits, these were compiled with a SJLJ exception mechanism.
+
+If one chooses a MinGW compilation, we need to respect these exception mechanism.
```
these -> this

chrisdembia commented on README.md:
```diff
@@ -466,6 +466,60 @@ Simbody is installed into the `usr/` directory.  The directory
+
+To see which exception mechanism is used, user can have a look at dlls located in the `bin` directory of MinGW.
+The name of mechanism is present in the file `libgcc_XXXX.dll`, where `XXXX` can be `dw`, `seh` or `sljl`.
+On some MinGW version, this information is also available by looking at the result of `gcc --version`.
```
"user can have a look" -> "the user can look"

"On some MinGW version" -> "For some MinGW versions"

chrisdembia commented on README.md:
```diff
@@ -466,6 +466,60 @@ Simbody is installed into the `usr/` directory.  The directory
+A check is made on the MinGW version used, and if the exception mechanism is different,
+then the configuration may stop because of this difference.
+If one provides a Blas and Lapack libraries with flag `BUILD_USING_OTHER_LAPACK`,
+compilation with MinGW is always possible.
```
"A check is made on the MinGW version used" -> "CMake will check the version of your MinGW"

"If one provides a Blas and Lapack libraries with flag" -> "If one provides Blas and Lapack libraries with the CMake variable"

chrisdembia commented on Platform/CMakeLists.txt:
```diff
@@ -19,6 +19,43 @@ if(WIN32 AND NOT CMAKE_HOST_SYSTEM_NAME MATCHES Windows)
+    if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+        set(ARCH64_BITS TRUE)
+    else()
+        set(ARCH64_BITS FALSE)
```
You can use the `PLATFORM_ABI` variable that is set in the root CMakeLists.txt. https://github.com/Gjacquenot/simbody/blob/master/CMakeLists.txt#L198

Gjacquenot commented:
@sherm1 : I have checked my version and downloaded a new version (link below), and I see that the DLLS needed are present twice in the `bin` directory and also in the `x86_64-w64-mingw32\lib\` directory. My bet that everything needed was in `bin` is not so good...

@sherm1  and @chrisdembia : I will investigate this and create a new pull request that takes into account the new commits and all remarks you have done.

http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/4.9.3/threads-posix/seh/x86_64-4.9.3-release-posix-seh-rt_v4-rev1.7z/download

sherm1 commented:
@Gjacquenot, you don't need to open a new pull request. You can just push more commits to the same branch you have been working on and they will automatically show up in this PR. A PR is a live connection between your branch and the one you are submitting to.

chrisdembia commented:
@Gjacquenot just make sure to add a comment to the PR when you've made all the changes. GitHub doesn't send out a notification when more commits are added to a PR.

Gjacquenot commented:
@sherm1 : I had never met a MinGW installation without DLL in the bin directory. I have understood that MinGW can do cross compiling and store DLL in two lib directories depending on the target choice. I now use the DLL located in the lib directory. I also added some checks in CMake
@chrisdembia : I have fixed the typos you mention

chrisdembia commented:
@Gjacquenot thanks for fixing those typos.

You could add yourself to the table of contributors in the CONTRIBUTING.md file.

sherm1 commented:
Hi, @Gjacquenot. Thanks for updating. I was able to run CMake successfully this time but ran into a problem -- when I typed "make" from the MSYS shell, I got the message `pthread.h: No such file or directory` (full output below). Are you expecting that MinGW provides that or should it be using the one Simbody provides? In case it matters, I have my source and binary directories separate, and I used the CMake 3.3.1 GUI to generate the makefiles, taking all defaults.

A lesser issue -- in the output below it appears that the libraries are being installed twice.
```
$ make
Scanning dependencies of target PlatformFiles_subdir
[  0%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\libblas.lib -> C;C:\msys\1.0\Users\Sh
erm\Documents\Binaries\VS14-64\simbody-master-msys\libblas.lib
[  0%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\liblapack.lib -> C;C:\msys\1.0\Users\
Sherm\Documents\Binaries\VS14-64\simbody-master-msys\liblapack.lib
[  0%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\pthreadVC2_x64.lib -> C;C:\msys\1.0\U
sers\Sherm\Documents\Binaries\VS14-64\simbody-master-msys\pthreadVC2_x64.lib
[  0%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\libblas.dll -> C;C:\msys\1.0\Users\Sh
erm\Documents\Binaries\VS14-64\simbody-master-msys\libblas.dll
[  0%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\libgcc_s_sjlj-1.dll -> C;C:\msys\1.0\
Users\Sherm\Documents\Binaries\VS14-64\simbody-master-msys\libgcc_s_sjlj-1.dll
[  1%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\libgfortran-3.dll -> C;C:\msys\1.0\Us
ers\Sherm\Documents\Binaries\VS14-64\simbody-master-msys\libgfortran-3.dll
[  1%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\liblapack.dll -> C;C:\msys\1.0\Users\
Sherm\Documents\Binaries\VS14-64\simbody-master-msys\liblapack.dll
[  1%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\libquadmath-0.dll -> C;C:\msys\1.0\Us
ers\Sherm\Documents\Binaries\VS14-64\simbody-master-msys\libquadmath-0.dll
[  1%] Copy C;C:\msys\1.0\Users\Sherm\Documents\GitHub\simbody\simbody-master\Platform\Windows\lib_x64\pthreadVC2_x64.dll -> C;C:\msys\1.0\U
sers\Sherm\Documents\Binaries\VS14-64\simbody-master-msys\pthreadVC2_x64.dll
[  1%] Built target PlatformFiles_subdir
Scanning dependencies of target PlatformFiles
[  1%] Built target PlatformFiles
Scanning dependencies of target SimTKcommon
[  1%] Building CXX object SimTKcommon/sharedTarget/CMakeFiles/SimTKcommon.dir/__/src/About.cpp.obj
[  1%] Building CXX object SimTKcommon/sharedTarget/CMakeFiles/SimTKcommon.dir/__/src/AtomicInteger.cpp.obj
In file included from c:/Users/Sherm/Documents/GitHub/simbody/simbody-master/SimTKcommon/src/AtomicInteger.cpp:25:0:
c:/Users/Sherm/Documents/GitHub/simbody/simbody-master/SimTKcommon/src/gmx_atomic.h:67:21: fatal error: pthread.h: No such file or directory

compilation terminated.
```

Gjacquenot commented:
Once exception problem has been taken care of, another problem arises...

Could you please tell me which version of MinGW you got? and maybe where you downloaded it?
I know that once you have selected the exception mechanism, one has to look at the thread mechanism. Are you using a `threads-win32` version of MinGW? I have tested the `threads-posix` and had no problem.
As I already tested the exception mechanism I can 

I plan to give a shell script to fetch and compile sources from Lapack and Simbody to test these installations. Maybe it could be stored as a seperate file or in a md file. What do you think?


It could look like this


    # 0 Path definition
    export CMakeGenerator="MSYS Makefiles"
    export Gcc="C:\MinGW64_520_Posix_SJLJ\bin\gcc.exe"
    export Gpp="C:\MinGW64_520_Posix_SJLJ\bin\g++.exe"
    export LapackInstallDirectory="C:\Program Files\Lapack_3_5_0__5_2_0"
    export SimbodyInstallDirectory="C:\Program Files\Simbody"
    
    mkdir simbodyWS
    cd simbodyWS
    
    # 1 Fetch sources
    wget http://www.netlib.org/lapack/lapack-3.5.0.tgz
    git clone https://github.com/simbody/simbody.git
    
    # 2 Build Lapack
    tar -xf lapack-3.5.0.tgz
    mkdir -p lapack_build
    cd lapack_build
    cmake ../lapack-3.5.0 \
        -G"${CMakeGenerator}" \
        -DCMAKE_C_COMPILER:PATH="${Gcc}" \
        -DCMAKE_CXX_COMPILER:PATH="${Gpp}" \
        -DCMAKE_BUILD_TYPE=Release \
        -DBUILD_COMPLEX:BOOL=ON \
        -DBUILD_COMPLEX16:BOOL=ON \
        -DBUILD_DOUBLE:BOOL=ON \
        -DBUILD_SINGLE:BOOL=ON \
        -DBUILD_SHARED_LIBS:BOOL=ON \
        -DBUILD_STATIC_LIBS:BOOL=OFF \
        -DLAPACKE:BOOL=OFF \
        -DLAPACKE_WITH_TMG:BOOL=OFF \
        -DCMAKE_INSTALL_PREFIX:PATH="${LapackInstallDirectory}"
    make
    make install
    cd ..
    
    # 3 Build Simbody
    mkdir -p simbody_build
    cd simbody_build
    cmake ../../simbody \
        -G"${CMakeGenerator}" \
        -DCMAKE_BUILD_TYPE=Release \
        -DCMAKE_INSTALL_PREFIX:PATH="${SimbodyInstallDirectory}" \
        -DCMAKE_C_COMPILER:PATH="${Gcc}" \
        -DCMAKE_CXX_COMPILER:PATH="${Gpp}" \
        -DBUILD_USING_OTHER_LAPACK:PATH="${LapackInstallDirectory}\bin\liblapack.dll;${LapackInstallDirectory}\bin\libblas.dll"
    make
    make test
    make install

chrisdembia commented:
If @sherm1 is fine with the idea of such a script, I think it should be stored in the simbody repo (in the root folder or in the doc folder); the README can link to it. It should assume that simbody has already been cloned (to obtain the script), and so cloning simbody shouldn't need to be in the script.

I also think it would be fine for the README to say "Can build with MinGW if you're uisng these specific MinGW downloads" or something like that; we don't need to support all possible MinGW installations (or those could be added later).

sherm1 commented:
> Could you please tell me which version of MinGW you got? and maybe where you downloaded it?

I don't know. I don't remember whether I got it from http://www.mingw.org/ or http://www.mingw-w64.org/. I don't remember seeing any options for choosing pthreads vs. win32 threads either. I would be happy to delete my installation entirely and reinstall following your instructions. I agree with Chris that we don't have to support every possible MinGW installation, but I think we have to be able to say "if you followed this procedure to install MinGW and MSYS, then you will be able to build Simbody". 

I think you should pick an approved way that MinGW/MSYS should be installed on Windows and then claim support for that and for nothing else. From my experience so far I would say that any untested installation type probably doesn't work. Maybe we can support whatever installation you have, and if you give me a different approach I can test that also. Do you have a way of testing different installations yourself?

BTW, I would be happy to stick with 64 bit only if that simplifies things.

Gjacquenot commented:
@sherm1 : I have downloaded many MinGW times from http://www.mingw.org. But now I use the forked version, which is more consistent with the versions produced.

For the 64 bits, I download version at this address:
http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/

For the 32 bits, I download version at this address:
http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/

I mostly used the threads-posix version.

I will test the thread win32 version, with the script proposed, that can be used to test a bunch of versions .

I will make a list of the version tested, with the corresponding download address.

Gjacquenot commented:
I have updated the CMakeLists.txt to allow only the posix thread model. The win32 thread model that @sherm1 has tested does not allow some thread functionality like `std:mutex`

I have tested 3 versions of MinGW with a posix thread model and various exception mechanism. 2 of 3 succeed

For one version, I still have some problem with opengl, that I have to investigate

    Scanning dependencies of target simbody-visualizer
    [ 64%] Building CXX object Simbody/Visualizer/simbody-visualizer/CMakeFiles/simbody-visualizer.dir/simbody-visualizer.cpp.obj
    In file included from D:\GJ\simbody\simbody\Simbody\Visualizer\simbody-visualizer\simbody-visualizer.cpp:55:0:
    D:\GJ\simbody\simbody\Simbody\Visualizer\simbody-visualizer\glut32/glext.h:7818:
    144: error: conflicting declaration 'typedef void (__attribute__((__stdcall__))
    * PFNGLCOLORSUBTABLEEXTPROC)(GLenum, GLsizei, GLsizei, GLenum, GLenum, const GLvoid*)'
     typedef void (APIENTRYP PFNGLCOLORSUBTABLEEXTPROC) (GLenum target, GLsizei start, GLsizei count, GLenum format, GLenum type, const GLvoid *data);
    
                                                                    ^
    In file included from D:\GJ\simbody\simbody\Simbody\Visualizer\simbody-visualizer\glut32/glut.h:148:0,
                     from D:\GJ\simbody\simbody\Simbody\Visualizer\simbody-visualizer\simbody-visualizer.cpp:54:
    C:/PROGRA~1/MINGW-~1/I686-5~1.0-P/mingw32/i686-w64-mingw32/include/GL/gl.h:1034:25: note: previous declaration as 'typedef void (* PFNGLCOLORSUBTABLEEXTPROC)(GL
    enum, GLsizei, GLsizei, GLenum, GLenum, const GLvoid*)'
     typedef void (APIENTRY *PFNGLCOLORSUBTABLEEXTPROC)(GLenum target,GLsizei start,GLsizei count,GLenum format,GLenum type,const GLvoid *data);
                             ^
    Simbody\Visualizer\simbody-visualizer\CMakeFiles\simbody-visualizer.dir\build.make:62: recipe for target 'Simbody/Visualizer/simbody-visualizer/CMakeFiles/simbody-visualizer.dir/simbody-visualizer.cpp.obj' failed
    mingw32-make[2]: *** [Simbody/Visualizer/simbody-visualizer/CMakeFiles/simbody-visualizer.dir/simbody-visualizer.cpp.obj] Error 1
    CMakeFiles\Makefile2:4832: recipe for target 'Simbody/Visualizer/simbody-visualizer/CMakeFiles/simbody-visualizer.dir/all' failed
    mingw32-make[1]: *** [Simbody/Visualizer/simbody-visualizer/CMakeFiles/simbody-visualizer.dir/all] Error 2
    makefile:137: recipe for target 'all' failed
    mingw32-make: *** [all] Error 2

sherm1 commented:
@Gjacquenot, I'm looking at this now. I keep having trouble updating my copy of your branch -- I get conflicts, but that shouldn't happen. I suspect that means you are doing rebases or cherry-picking -- is that right? You should never do that after anyone else has started using your commits.

chrisdembia commented:
> cherry-picking

I think cherry-picking doesn't destroy history; I think it makes a new commit with the same changes as another.

chrisdembia commented:
@sherm1 I think all will be okay if you first delete your local copy of @Gjacquenot's branch.

sherm1 commented:
> I think all will be okay if you first delete your local copy of @Gjacquenot's branch.

Thanks, Chris -- that's what I did to fix the problem. Not a big deal but it would be easier just to be able to pull the changes!

sherm1 commented:
> I think cherry-picking doesn't destroy history; I think it makes a new commit with the same changes as another.

Yes, that's right. But any way of getting changes in other than updating from the existing changes can make it hard for reviewers to pull the updates.

sherm1 commented:
I nuked my old mingw/msys installations and reinstalled mingw-w64 using the links you provided. 
- First I tried mingw64 pthreads/seh and verified that I got an appropriate error message in CMake.
- Then I removed that and install mingw64 pthreads/sjlj (5.2.0 version).
- I installed in c:/mingw64 and added c:/mingw64/bin to my Windows PATH.
- I ran CMake 3.3.1 and specified mingw make files as the generator.
- I built using mingw32-make per your instructions in the readme, and successfully ran all the tests and verified that an example using the visualizer worked correctly.
- I tried "mingw32-make install" and it seemed to install properly (I looked at the installation but didn't test it)
- I then tried using Visual Studio 2015 Win64 as the generator and verified that that still works correctly.

Some minor issues:
- I got two warnings during the build about `NOMINMAX` being redefined, in Simbody's [Pathname.cpp:39](https://github.com/simbody/simbody/blob/master/SimTKcommon/src/Pathname.cpp#L39) and [Plugin.cpp:39](https://github.com/simbody/simbody/blob/master/SimTKcommon/src/Plugin.cpp#L39). Those should have `#ifndef NOMINMAX` guards to avoid the warnings.
- The README instructions should mention that the mingw bin has to be on the path.
- Although I built 64 bit Simbody, the default installation directory in the CMake GUI was `Program Files (x86)` which is where 32 bit binaries should go. With the Visual Studio generator the CMake script is able to figure out whether to install in `Program Files` or `Program Files (x86)`; not sure why that isn't working here.

This is looking very good, Guillaume! Can you fix these minor things?

sherm1 commented on README.md:
```diff
@@ -466,6 +466,103 @@ Simbody is installed into the `usr/` directory.  The directory
+Otherwise, we need to provide our own Blas and Lapack libraries.
+
+To see which exception mechanism is used, user can look at dlls located in the `bin` directory of MinGW.
+The name of mechanism is present in the file `libgcc_XXXX.dll`, where `XXXX` can be `dw`, `seh` or `sljl`.
```
typo: sljl -> sjlj

Gjacquenot commented:
@sherm1 : i have done a rebase operation to fetch the new commits from simbody master branch.
I have done it so that I follow the evolution of master and fix conflicts quickly.

I am currently writing script to automate the test of mingw version. I want to do it with shell script and use cmake msys generator. Not mingw one. 
@sherm1 and @chrisdembia are you familiar with msys?

chrisdembia commented:
I'm not familiar with msys. Are you planning to include the shell script with this PR or a separate PR?

sherm1 commented on README.md:
```diff
@@ -466,6 +466,103 @@ Simbody is installed into the `usr/` directory.  The directory
+64 Bits Version | Posix  | SEH       | YES         | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.2.0/threads-posix/sjlj/x86_64-5.2.0-release-posix-sjlj-rt_v4-rev0.7z/download)    |
+64 Bits Version | Posix  | SJLJ      |             | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.2.0/threads-posix/seh/x86_64-5.2.0-release-posix-seh-rt_v4-rev0.7z/download)      |
+
+For the recommended versions, user does not have to provide its own Blas and Lapack libraries. In the other case, Blas and Lapack have to be compiled manually.
```
The table says the recommended version is SEH. But then the user will have to provide their own Blas/Lapack so either we should recommend SJLJ or change this sentence.

It does appear that it would be better if Simbody were using SEH and provided SEH libraries. However, I think that should be the topic of a separate issue now and a later PR.

sherm1 commented on README.md:
```diff
@@ -466,6 +466,103 @@ Simbody is installed into the `usr/` directory.  The directory
+The name of mechanism is present in the file `libgcc_XXXX.dll`, where `XXXX` can be `dw`, `seh` or `sljl`.
+For some MinGW version, this information is also available by looking at the result of `gcc --version`.
+
+CMake will check the version of your MinGW, and if the exception mechanism is different,
```
Should mention that mingw bin directory must be on the PATH for things to work smoothly (assuming that's true).

sherm1 commented:
> i have done a rebase operation to fetch the new commits from simbody master branch. I have done it so that I follow the evolution of master and fix conflicts quickly.

Yes, that's what I thought. It makes sense to do that when you are working privately but once you make your commits available to others (for example, in a PR) it is better to merge from master rather than rebase. In my copy of your branch I had your old commits that I had tested. When I tried to pull from you branch, your all-new rebased commits conflicted with the ones I had gotten before. If you had merged instead then my pull would have done exactly what you did and saved me some time and confusion.

sherm1 commented:
> @sherm1 and @chrisdembia are you familiar with msys?

I am. But in this case I deleted my msys installation and just followed your mingw instructions. I would be happy with just that for this PR; you could add msys capability in a later PR if you want.

sherm1 commented on README.md:
```diff
@@ -466,6 +466,103 @@ Simbody is installed into the `usr/` directory.  The directory
+
+#### Supported MinGW versions
+
+The table below lists the various versions of MinGW versions supported
```
I really like this table -- great idea!

Gjacquenot commented:
@sherm1 Sorry for that. I will merge next time.

MSys is not enough standard to be used here. I will stick to mingw generator and batch script. I will be missing a wget command to download mingw. I was thinking using a first python script to download a mingw version and eventually lapack source. (I don't know a wget programming shipped directly with Windows)

sherm1 commented:
> MSys is not enough standard to be used here. I will stick to mingw generator and batch script. 

I think simpler is better. Once you showed me where to get good binaries I didn't have any trouble following your instructions. Once this PR is merged users will be able to build in mingw by following the instructions, or if they have trouble they can post to the forum for guidance. It will be more difficult if they don't have a good lapack/blas binary though!

Gjacquenot commented:
I have succeeded in generating a batch script that a MinGW version and compiles simbody with the blas and lapack libraries. Another one is used to test simbody with user defined version of blas and lapack

With the `i686-5.2.0-posix-sjlj-rt_v4-rev0` version , I still have a problem with compiling the visualizer. I could use a compiled version of freeglut but I want to stick with what is provided here.

I have a prototype conflict in file `Simbody\Visualizer\simbody-visualizer\simbody-visualizer.cpp:54`

    D:\GJ\simbody\simbody\Simbody\Visualizer\simbody-visualizer\glut32/glext.h:7818:
    144: error: conflicting declaration 'typedef void (__attribute__((__stdcall__))* PFNGLCOLORSUBTABLEEXTPROC)(GLenum, GLsizei, GLsizei, GLenum, GLenum, const GLvoid*)'
     typedef void (APIENTRYP PFNGLCOLORSUBTABLEEXTPROC) (GLenum target, GLsizei start, GLsizei count, GLenum format, GLenum type, const GLvoid *data);

    In file included from D:\GJ\simbody\simbody\Simbody\Visualizer\simbody-visualizer\glut32/glut.h:148:0,
    C:/PROGRA~1/MINGW-~1/I686-5~1.0-P/mingw32/i686-w64-mingw32/include/GL/gl.h:1034:
    25: note: previous declaration as 'typedef void (* PFNGLCOLORSUBTABLEEXTPROC)(GLenum, GLsizei, GLsizei, GLenum, GLenum, const GLvoid*)'
     typedef void (APIENTRY *PFNGLCOLORSUBTABLEEXTPROC)(GLenum target,GLsizei start,GLsizei count,GLenum format,GLenum type,const GLvoid *data);

From what I understood, it is a convention call problem with `__stdcall__`

Gjacquenot commented:
I am investigating the compilation problem with the visualizer, where one links against compiled glut libraries.

It definitely seems to be a problem of calling a MSVC library in a MinGW environnement.

DependencyWalker seems to indicate the following conventions are used:

    __stdcall : simbody\visualizer\simbody-visualizer\glut32\lib\GLUT32.DLL

The MinGW version for 32 bits seems to be using `__cdecl` declaration, which is why the compilation crashes.

I do not meet the problem with the 64 bit version, since calling conventions seems to match.

    __cdecl   : simbody\visualizer\simbody-visualizer\glut32\lib64\GLUT32.DLL

Solving this problem will require some time...

@sherm1 and @chrisdembia : For the time being, I would like to add an input flag to disable visualization `-DDISABLE_VISUALIZER`

    if(DEFINED DISABLE_VISUALIZER)
        set(OPENGL_FOUND 0)
        set(SIMBODY_HAS_GLUT FALSE)
        return()
    endif()

To summarize, here is the status for compiling Simbody with MinGW

OS              | Thread | Exception | Recommended | Success  | URL                                                                                                                                                                                                                            |
----------------|--------|-----------|------------ |----------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
32 Bits Version | Posix  | Dwarf     | YES         |          | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/5.2.0/threads-posix/dwarf/i686-5.2.0-release-posix-dwarf-rt_v4-rev0.7z/download)    |
32 Bits Version | Posix  | SJLJ      |             |          | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/5.2.0/threads-posix/sjlj/i686-5.2.0-release-posix-sjlj-rt_v4-rev0.7z/download)      |
64 Bits Version | Posix  | SJLJ      | YES         | OK       | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.2.0/threads-posix/sjlj/x86_64-5.2.0-release-posix-sjlj-rt_v4-rev0.7z/download)    |
64 Bits Version | Posix  | SEH       |             | OK       | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.2.0/threads-posix/seh/x86_64-5.2.0-release-posix-seh-rt_v4-rev0.7z/download)      |

sherm1 commented:
> I would like to add an input flag to disable visualization -DDISABLE_VISUALIZER

There is already a CMake option variable `BUILD_VISUALIZER` in the Simbody CMakeLists.txt. It defaults to "on" but we turn it off for our Travis CI builds. Can you use that?

Gjacquenot commented:
Thanks @sherm1, by adding  `-DBUILD_VISUALIZER:BOOL=OFF` on the cmake command, compilation succeeds on 32 Bits platform.

Compilation status:

OS      | Thread | Exception | Comment                                                             | URL                                                                                                                                                                                                                            |
--------|--------|-----------|---------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
64 Bits | Posix  | SJLJ      | All features supported, all binary included                         | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.2.0/threads-posix/sjlj/x86_64-5.2.0-release-posix-sjlj-rt_v4-rev0.7z/download)    |
64 Bits | Posix  | SEH       | Needs to be linked against user's Blas and Lapack                   | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.2.0/threads-posix/seh/x86_64-5.2.0-release-posix-seh-rt_v4-rev0.7z/download)      |
32 Bits | Posix  | Dwarf     | No visualization, all binary included                               | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/5.2.0/threads-posix/dwarf/i686-5.2.0-release-posix-dwarf-rt_v4-rev0.7z/download)    |
32 Bits | Posix  | SJLJ      | No visualization, needs to be linked against user's Blas and Lapack | [MinGW64 project GCC 5.2.0](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/5.2.0/threads-posix/sjlj/i686-5.2.0-release-posix-sjlj-rt_v4-rev0.7z/download)      |

sherm1 commented:
So is this ready to merge then, Guillaume?

Gjacquenot commented:
Just 2 more commits to come:
- one to disable visualize on MinGW 32 bit
- one to update the Readme.md

Gjacquenot commented:
I think it is now ok!

sherm1 commented:
- the new table looks great in your comment above, but it didn't seem to format correctly when I viewed README.md in GitHub
- I retried the 64 bit build on my machine with no trouble, but noticed that the default installation location is still `/Program Files (x86)/Simbody` which is the 32 bit installation location. I'm OK with leaving it this way if you want since your instructions specify a particular installation directory but I wanted to mention it in case you meant to fix it.

Gjacquenot commented:
I guess these two last commits complete the MinGW integration.

I can start working with SimBody!!

The rest of this post has nothing to do with MinGW configuration and compilation.

I have just met a weird behaviour with `Vec3`. The default constructor does not behave like other mathematical object like `Rotation`. This is weird with the following code.


    #include "SimTKsimbody.h"
    using namespace SimTK;
    using namespace std;
    int main()
    {
        cout<<Transform(Rotation(),Vec3())<<endl;
        cout<<Transform(Rotation(),Vec3(0.0))<<endl;
        return 0;
    }

which gives

    [1,0,0,nan]
    [0,1,0,nan]
    [0,0,1,nan]
    [0,0,0,1]
    
    [1,0,0,0]
    [0,1,0,0]
    [0,0,1,0]
    [0,0,0,1]

I have found this *feature*, while understanding that two colliding objects did not give contact efforts.
I would have expected that some *function*, *method* throws an exception to tell me that my transform matrix had *nan*.

I can understand the `Vec3` default constructor, but it would be nice to check transform matrix when using it.

sherm1 commented:
Everything looks good now -- I'll merge this.

> I can understand the Vec3 default constructor, but it would be nice to check transform matrix when using it.

Good catch -- I agree. It should check in Debug builds. The `Vec3` default constructor intentionally leaves memory uninitialized (NaN in Debug) because of the memory-touching waste that would occur in large aggregates like `Vector_<Vec3>` if they were unnecessarily initialized to zero. But a Rotation and Transform are not supposed to be uninitialized.

Please open a new issue and transfer your very helpful example and comment there; this should be separate from the MinGW build PR.

Thanks very much for your excellent MinGW contribution!
